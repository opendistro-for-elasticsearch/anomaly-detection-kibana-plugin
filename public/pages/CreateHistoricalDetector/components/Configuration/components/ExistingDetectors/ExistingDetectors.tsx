/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import {
  EuiFlexItem,
  EuiFlexGroup,
  EuiSpacer,
  EuiComboBox,
  EuiRadioGroup,
  EuiCallOut,
} from '@elastic/eui';
import { FormikProps } from 'formik';
import { debounce, get } from 'lodash';
import React, { useState, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { AppState } from '../../../../../../redux/reducers';
import {
  getDetectorList,
  getHistoricalDetectorList,
} from '../../../../../../redux/reducers/ad';
import { getMappings } from '../../../../../../redux/reducers/elasticsearch';
import { GET_ALL_DETECTORS_QUERY_PARAMS } from '../../../../../utils/constants';
import { searchDetector } from '../../../../../../redux/reducers/ad';
import { FormattedFormRow } from '../../../../../createDetector/components/FormattedFormRow/FormattedFormRow';
import { HistoricalDetectorFormikValues } from '../../../../utils/constants';
import { CREATE_HISTORICAL_DETECTOR_OPTIONS } from '../../../../utils/constants';
import { sanitizeSearchText } from '../../../../../utils/helpers';
import {
  populateDetectorFieldsFromDetector,
  populateDetectorFieldsToInitialValues,
  untouchDetectorFields,
} from './utils/helpers';
import { getAllDetectorOptions } from '../../../../utils/helpers';
import { GetDetectorsQueryParams } from '../../../../../../../server/models/types';

interface ExistingDetectorsProps {
  formikProps: FormikProps<HistoricalDetectorFormikValues>;
}

export function ExistingDetectors(props: ExistingDetectorsProps) {
  const dispatch = useDispatch();
  const selectedIndexOption = get(
    props,
    'formikProps.values.index.0.label',
    ''
  );
  const isIndexSelected = selectedIndexOption && selectedIndexOption.length > 0;
  const adState = useSelector((state: AppState) => state.ad);
  const detectors = adState.detectors;
  const allDetectors = {
    ...adState.detectorList,
    ...adState.historicalDetectorList,
  };
  const allDetectorOptions = isIndexSelected
    ? getAllDetectorOptions(Object.values(allDetectors))
    : [];
  const [queryText, setQueryText] = useState('');
  const [selectedDetectorId, setSelectedDetectorId] = useState<string>('');
  const selectedDetector = detectors[selectedDetectorId];
  const [createDetectorSelection, setCreateDetectorSelection] = useState<
    CREATE_HISTORICAL_DETECTOR_OPTIONS
  >(CREATE_HISTORICAL_DETECTOR_OPTIONS.CREATE_NEW);

  const existingDetectorsOptions = [
    {
      id: CREATE_HISTORICAL_DETECTOR_OPTIONS.CREATE_NEW,
      label: 'No',
    },
    {
      id: CREATE_HISTORICAL_DETECTOR_OPTIONS.USE_EXISTING,
      label: 'Yes',
    },
  ];

  const fetchAllDetectors = async (params: GetDetectorsQueryParams) => {
    dispatch(getDetectorList(params));
    dispatch(getHistoricalDetectorList(params));
  };

  // Fetch all realtime and historical detectors that are using the specified index
  useEffect(() => {
    if (isIndexSelected) {
      const params = {
        ...GET_ALL_DETECTORS_QUERY_PARAMS,
        indices: selectedIndexOption,
      };
      fetchAllDetectors(params);
    }
  }, [selectedIndexOption]);

  // Update the form if a change in selected detector, and get index mappings
  useEffect(() => {
    if (selectedDetector) {
      populateDetectorFieldsFromDetector(props.formikProps, selectedDetector);
      const indexName = selectedDetector.indices[0];
      dispatch(getMappings(indexName));
    } else {
      populateDetectorFieldsToInitialValues(props.formikProps);
    }
  }, [selectedDetector]);

  const getDetectorInfo = (detectorId: string) => {
    if (detectorId && allDetectors && allDetectors[detectorId]) {
      dispatch(
        searchDetector({
          query: { term: { 'name.keyword': allDetectors[detectorId].name } },
        })
      );
    }
  };

  const handleSearchChange = debounce(async (searchValue: string) => {
    if (searchValue !== queryText) {
      const sanitizedQuery = sanitizeSearchText(searchValue);
      setQueryText(sanitizedQuery);
      if (isIndexSelected) {
        const params = {
          ...GET_ALL_DETECTORS_QUERY_PARAMS,
          search: sanitizedQuery,
          indices: selectedIndexOption,
        };
        fetchAllDetectors(params);
      }
    }
  }, 300);

  const onSelectionChange = (id: string) => {
    setCreateDetectorSelection(id as CREATE_HISTORICAL_DETECTOR_OPTIONS);
    if (id === CREATE_HISTORICAL_DETECTOR_OPTIONS.CREATE_NEW) {
      setSelectedDetectorId('');
    }
    populateDetectorFieldsToInitialValues(props.formikProps);
    untouchDetectorFields(props.formikProps);
  };

  return (
    <EuiFlexGroup direction="column" style={{ margin: '0px' }}>
      <EuiFlexGroup
        direction="row"
        gutterSize="none"
        justifyContent="spaceBetween"
        style={{ maxWidth: '400px' }}
      >
        <EuiFlexItem>
          <FormattedFormRow
            title="Configure using an existing detector"
            hint="Choose if you would like to source an existing real-time or historical detector to use as a template for your configuration."
          >
            <EuiRadioGroup
              name="use existing detectors radio group"
              options={existingDetectorsOptions}
              idSelected={createDetectorSelection}
              onChange={onSelectionChange}
            />
          </FormattedFormRow>
        </EuiFlexItem>
      </EuiFlexGroup>
      <EuiSpacer size="l" />
      {createDetectorSelection ===
      CREATE_HISTORICAL_DETECTOR_OPTIONS.USE_EXISTING ? (
        <EuiFlexGroup direction="column">
          {!isIndexSelected ? (
            <EuiCallOut
              style={{ marginLeft: '12px' }}
              title="No index has been selected. Please select an index first."
              color="warning"
              iconType="alert"
              size="s"
            />
          ) : allDetectorOptions.length === 0 ? (
            <EuiCallOut
              style={{ marginLeft: '12px' }}
              title="No existing detectors are using the selected index."
              color="warning"
              iconType="alert"
              size="s"
            />
          ) : null}
          <EuiFlexItem style={{ maxWidth: '70%' }}>
            <FormattedFormRow
              fullWidth
              title="Source existing detector configuration"
              hint="Choose from an existing real-time or historical detector configuration"
            >
              <EuiComboBox
                id="existingDetectorsComboBox"
                placeholder="Existing detector configurations"
                async
                isLoading={adState.requesting}
                isDisabled={allDetectorOptions.length === 0}
                options={allDetectorOptions}
                onSearchChange={handleSearchChange}
                onBlur={() => {}}
                onChange={(options) => {
                  const detectorId = get(options, '0.id') as string;
                  setSelectedDetectorId(detectorId);
                  getDetectorInfo(detectorId);
                }}
                selectedOptions={
                  (selectedDetectorId &&
                    allDetectors &&
                    allDetectors[selectedDetectorId] && [
                      { label: allDetectors[selectedDetectorId].name },
                    ]) ||
                  []
                }
                singleSelection={true}
                isClearable={true}
              />
            </FormattedFormRow>
          </EuiFlexItem>
          <EuiSpacer />
        </EuiFlexGroup>
      ) : null}

      {createDetectorSelection ===
        CREATE_HISTORICAL_DETECTOR_OPTIONS.USE_EXISTING &&
      selectedDetector !== undefined ? (
        <EuiSpacer size="l" />
      ) : null}
    </EuiFlexGroup>
  );
}
