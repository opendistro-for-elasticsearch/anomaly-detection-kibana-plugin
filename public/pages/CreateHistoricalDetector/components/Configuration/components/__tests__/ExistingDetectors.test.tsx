/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React from 'react';
import { Provider } from 'react-redux';
import {
  HashRouter as Router,
  RouteComponentProps,
  Route,
  Switch,
} from 'react-router-dom';
import { render, fireEvent } from '@testing-library/react';
import { ExistingDetectors } from '../ExistingDetectors/ExistingDetectors';
import configureStore from '../../../../../../redux/configureStore';
import {
  httpClientMock,
  coreServicesMock,
} from '../../../../../../../test/mocks';
import { CoreServicesContext } from '../../../../../../components/CoreServices/CoreServices';
import { Formik } from 'formik';

const TITLE_TEXT = 'Configure using an existing detector';
const HELP_TEXT =
  'Choose if you would like to source an existing real-time or historical detector to use as a template for your configuration.';
const NO_INDEX_CALLOUT_TEXT =
  'No index has been selected. Please select an index first.';
const NO_DETECTORS_CALLOUT_TEXT =
  'No existing detectors are using the selected index.';

const renderWithRouter = (isEdit: boolean = false, formikValues: {} = {}) => ({
  ...render(
    <Provider store={configureStore(httpClientMock)}>
      <Router>
        <Switch>
          <Route
            render={(props: RouteComponentProps) => (
              <CoreServicesContext.Provider value={coreServicesMock}>
                <Formik
                  initialValues={{ index: [{ label: 'test-index' }] }}
                  onSubmit={jest.fn()}
                >
                  {() => (
                    <div>
                      <ExistingDetectors
                        formikProps={{
                          values: formikValues,
                          setFieldValue: jest.fn(),
                          setFieldTouched: jest.fn(),
                        }}
                        {...props}
                      />
                    </div>
                  )}
                </Formik>
              </CoreServicesContext.Provider>
            )}
          />
        </Switch>
      </Router>
    </Provider>
  ),
});

describe('<ExistingDetectors /> spec', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });
  test('renders the component', () => {
    const { container, getByText } = renderWithRouter(false);
    expect(container.firstChild).toMatchSnapshot();
    getByText(TITLE_TEXT);
    getByText(HELP_TEXT);
    getByText('No');
    getByText('Yes');
  });
  test('renders no index callout if applicable', () => {
    const { getByText, queryByText } = renderWithRouter(false, {});
    getByText(TITLE_TEXT);
    getByText(HELP_TEXT);
    expect(queryByText(NO_INDEX_CALLOUT_TEXT)).toBeNull();
    fireEvent.click(getByText('Yes'));
    getByText(NO_INDEX_CALLOUT_TEXT);
  });
  test('renders no eligible detectors callout if applicable', () => {
    const propsWithIndex = { index: [{ label: 'test-index' }] };
    const { getByText, queryByText } = renderWithRouter(false, propsWithIndex);
    getByText(TITLE_TEXT);
    getByText(HELP_TEXT);
    expect(queryByText(NO_DETECTORS_CALLOUT_TEXT)).toBeNull();
    fireEvent.click(getByText('Yes'));
    getByText(NO_DETECTORS_CALLOUT_TEXT);
  });
});
