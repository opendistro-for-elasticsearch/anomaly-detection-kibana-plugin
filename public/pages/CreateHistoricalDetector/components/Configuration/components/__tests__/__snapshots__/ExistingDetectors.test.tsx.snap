// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ExistingDetectors /> spec renders the component 1`] = `
<div>
  <div
    class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionColumn euiFlexGroup--responsive"
    style="margin: 0px;"
  >
    <div
      class="euiFlexGroup euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
      style="max-width: 400px;"
    >
      <div
        class="euiFlexItem"
      >
        <div
          class="euiFormRow"
          hint="Choose if you would like to source an existing real-time or historical detector to use as a template for your configuration."
          id="random_html_id-row"
          title="Configure using an existing detector"
        >
          <div
            class="euiFormRow__labelWrapper"
          >
            <label
              class="euiFormLabel euiFormRow__label"
              for="random_html_id"
            >
              <div
                style="line-height: 8px;"
              >
                <p>
                  Configure using an existing detector
                </p>
                <br />
                <p
                  class="sublabel"
                >
                  Choose if you would like to source an existing real-time or historical detector to use as a template for your configuration.
                </p>
              </div>
            </label>
          </div>
          <div
            class="euiFormRow__fieldWrapper"
          >
            <div
              id="random_html_id"
            >
              <div
                class="euiRadio euiRadioGroup__item"
              >
                <input
                  checked=""
                  class="euiRadio__input"
                  id="create_new_detector"
                  name="use existing detectors radio group"
                  type="radio"
                  value=""
                />
                <div
                  class="euiRadio__circle"
                />
                <label
                  class="euiRadio__label"
                  for="create_new_detector"
                >
                  No
                </label>
              </div>
              <div
                class="euiRadio euiRadioGroup__item"
              >
                <input
                  class="euiRadio__input"
                  id="use_existing_detector"
                  name="use existing detectors radio group"
                  type="radio"
                  value=""
                />
                <div
                  class="euiRadio__circle"
                />
                <label
                  class="euiRadio__label"
                  for="use_existing_detector"
                >
                  Yes
                </label>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="euiSpacer euiSpacer--l"
    />
  </div>
</div>
`;
